backend/src/datasource.js

```diff
@@ -0,0 +1,67 @@
+import { DataSource } from "apollo-datasource";
+import crypto from "crypto";
+
+export class Post {
+    constructor(data) {
+        this.id = crypto.randomBytes(16).toString("hex");
+        this.voters = new Map();
+        Object.assign(this, data);
+    }
+}
+export class User {
+    constructor(name) {
+        this.name = name;
+    }
+}
+
+export class InMemoryDataSource extends DataSource {
+    constructor() {
+        super();
+        this.posts = [];
+        this.users = [];
+    }
+
+    initialize({ context }) {}
+
+    allUsers() {
+        return Promise.resolve(this.users);
+    }
+
+    getUser(name) {
+        return Promise.resolve(this.users.find((user) => user.name === name));
```

Comment:
Similar to calculating the number of votes on each requests, using `find` here (and in the other methods) has a worst-case complexity of O(n), though - as pretty much always in this assignment - there won't be any noticeable performance issue with this implementation.

Conceptually, `this.posts` and `this.users` are used to as lookup-tables and such could be also modelled as `Map` objects. This naturally prevents duplicates from being introduced, too.